File: 2023/01/01.py
import re
s1, s2 = 0, 0
digits = "0|1|2|3|4|5|6|7|8|9|_|one|two|three|four|five|six|seven|eight|nine"
as_num = lambda num: digits.split('|').index(num) % 10

for line in open(0):
    nums1 = [as_num(n[1]) for n in re.finditer(f'(?=({digits[:20]}))', line) if n[1]]
    nums2 = [as_num(n[1]) for n in re.finditer(f'(?=({digits}))', line) if n[1]]
    s1 += nums1[0]*10 + nums1[-1] if nums1 else 0
    s2 += nums2[0]*10 + nums2[-1] if nums2 else 0
print(s1, s2, sep="\n")

File: 2023/01/01.rs
use std::io::Read;

const DIGITS: [&str; 10] = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
const WORDS: [&str; 10] = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

fn main() {
    let mut input = String::new();
    std::io::stdin().read_to_string(&mut input).unwrap();

    let mut s1 = 0;
    let mut s2 = 0;
    for line in input.lines() {
        let mut nums: Vec<(usize, i32)> = vec![];
        DIGITS.iter().for_each(|digit| {
            line.match_indices(digit)
                .for_each(|i| nums.push((i.0, i.1.parse().unwrap())))
        });
        nums.sort();
        s1 += nums.first().unwrap_or(&(0, 0)).1 * 10 + nums.last().unwrap_or(&(0, 0)).1;

        WORDS.iter().for_each(|digit| {
            line.match_indices(digit).for_each(|i| {
                nums.push((i.0, WORDS.iter().position(|&x| x == i.1).unwrap() as i32))
            })
        });
        nums.sort();
        s2 += nums.first().unwrap_or(&(0, 0)).1 * 10 + nums.last().unwrap_or(&(0, 0)).1;
    }
    println!("{}\n{}", s1, s2);
}